### ISSUE TITLE
CharacterErrorRate drops checkpoint state and fails on empty distributed ranks

### ISSUE THREAD

**Issue by aswanth-07** [pytorch/ignite#3828]
### Bug

`CharacterErrorRate` does not fully participate in Ignite's metric state and distributed-reduction protocols:

1. It does not define `_state_dict_all_req_keys`, so `state_dict()` contains none of its accumulators and loading that state loses all progress.
2. `compute()` reduces `_num_errors` and `_num_refs`, but not `_num_examples`. If one distributed rank receives no local batches while another rank receives valid data, the empty rank still sees `_num_examples == 0` after the other accumulators have been reduced and raises `NotComputableError` instead of returning the global CER.

### Reproduction

The state loss is visible without a distributed setup:

```python
from ignite.metrics.nlp import CharacterErrorRate

metric = CharacterErrorRate()
metric.update((["helo"], ["hello"]))

restored = CharacterErrorRate()
restored.load_state_dict(metric.state_dict())
print(restored.compute())
```

On current `master` (`de08e79`), `restored.compute()` raises `NotComputableError`; it should return `0.2`.

The distributed case occurs with uneven input shards: update the metric only on rank 0, then call `compute()` collectively. Other ranks raise because `_num_examples` remains local even though errors and reference characters are globally reduced.

### Proposed fix

Declare all three accumulators in `_state_dict_all_req_keys` and include `_num_examples` in `sync_all_reduce`, matching the bookkeeping used by Ignite's other accumulating metrics. I have focused state round-trip and empty-rank distributed regression tests ready.

**Comment by vfdev-5** [comment]
> If one distributed rank receives no local batches while another rank receives valid data, the empty rank still sees _num_examples == 0 after the other accumulators have been reduced and raises NotComputableError instead of returning the global CER.

@aswanth-07 do you have a real usecase of that or some AI agent hinted you that ?

**Comment by aswanth-07** [comment]
Good question. I do not have a production incident for the uneven-shard case. I found the bookkeeping mismatch during a repository audit with AI assistance, as disclosed in the PR. A concrete case is distributed evaluation over an iterable or filtered dataset where one rank yields no batch while another rank does, but every rank must still enter compute() for the collective reductions. The checkpoint state loss is independently reproducible. If Ignite does not intend to support uneven per-rank metric updates, I am happy to narrow the issue and PR to the checkpoint fix.

**Comment by vfdev-5** [comment]
> A concrete case is distributed evaluation over an iterable or filtered dataset where one rank yields no batch while another rank does, but every rank must still enter compute() for the collective reductions. 

This case is possible when iterable dataset is sharded unevenly, but in this case metric update() method will still be called by all ranks for some number and then the remaining part can be call with a subset of ranks. So, compute wont raise the error you stated. 
And even in your particular case when just one rank calls update method, all reduce should still fix the problem...

**Comment by aswanth-07** [comment]
Thanks for clarifying. Yes, the all-reduce should fix it, and that is the change in #3829. Before the patch, CharacterErrorRate reduces _num_errors and _num_refs, but omits _num_examples, then checks the still-local _num_examples. The failure only occurs when a rank receives zero updates for the entire evaluation, for example when an evaluated or filtered shard has no valid examples. If every rank receives at least one batch, the existing check does not fail. I agree this is a narrow edge case, so I can drop the distributed part and keep only the checkpoint fix if zero-update ranks are outside Ignite's intended behavior.
